cmd/csi-resizer/main.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"time"
+
+	"github.com/kubernetes-csi/external-resizer/pkg/controller"
+	"github.com/kubernetes-csi/external-resizer/pkg/resizer"
+	"github.com/kubernetes-csi/external-resizer/pkg/util"
+
+	"k8s.io/klog"
+)
+
+var (
+	master       = flag.String("master", "", "Master URL to build a client config from. Either this or kubeconfig needs to be set if the provisioner is being run out of cluster.")
+	kubeConfig   = flag.String("kubeconfig", "", "Absolute path to the kubeconfig")
+	resyncPeriod = flag.Duration("resync-period", time.Minute*10, "Resync period for cache")
+	workers      = flag.Int("workers", 10, "Concurrency to process multiple resize requests")
+
+	enableLeaderElection      = flag.Bool("leader-election", false, "Enable leader election.")
+	leaderElectionIdentity    = flag.String("leader-election-identity", "", "Unique identity of this resizer. Typically name of the pod where the resizer runs.")
+	leaderElectionNamespace   = flag.String("leader-election-namespace", "kube-system", "Namespace where this resizer runs.")
+	leaderElectionRetryPeriod = flag.Duration("leader-election-retry-period", time.Second*5,
+		"The duration the clients should wait between attempting acquisition and renewal "+
+			"of a leadership. This is only applicable if leader election is enabled.")
+	leaderElectionLeaseDuration = flag.Duration("leader-election-lease-duration", time.Second*15,
+		"The duration that non-leader candidates will wait after observing a leadership "+
+			"renewal until attempting to acquire leadership of a led but unrenewed leader "+
+			"slot. This is effectively the maximum duration that a leader can be stopped "+
+			"before it is replaced by another candidate. This is only applicable if leader "+
+			"election is enabled.")
+	leaderElectionRenewDeadLine = flag.Duration("leader-election-renew-deadline", time.Second*10,
+		"The duration that non-leader candidates will wait after observing a leadership "+
+			"renewal until attempting to acquire leadership of a led but unrenewed leader "+
+			"slot. This is effectively the maximum duration that a leader can be stopped "+
+			"before it is replaced by another candidate. This is only applicable if leader "+
+			"election is enabled.")
+)
```

Comment:
Is there any known pattern for tunable patterns to expose for leader election? The external-attacher seems to expose far smaller number of tunables. cc @msau42

Reply:
External-attacher doesn't expose these three durations: `leaderElectionRetryPeriod`, `leaderElectionLeaseDuration`, and `leaderElectionRenewDeadLine`, it just use the default values:
https://github.com/kubernetes-csi/external-attacher/blob/master/cmd/csi-attacher/leader.go#L37

I make these args configurable because the core kube-controller-manager does so: https://github.com/kubernetes/kubernetes/blob/35b17dd6d08589bc6182d8038be1cf86d03c6ab0/cmd/controller-manager/app/options/generic.go#L81. However, users do not need to set these args in most cases, because we have provided appropriate default values.

I'm not sure which one is better, do we have any restrictions or rules for leader election?